src/coreclr/nativeaot/Runtime/windows/PalRedhawkMinWin.cpp

```diff
@@ -53,15 +53,49 @@ static uint32_t g_flsIndex = FLS_OUT_OF_INDEXES;
 void __stdcall FiberDetachCallback(void* lpFlsData)
 {
     ASSERT(g_flsIndex != FLS_OUT_OF_INDEXES);
+    ASSERT(g_flsIndex != NULL);
     ASSERT(lpFlsData == FlsGetValue(g_flsIndex));
 
-    if (lpFlsData != NULL)
+    // The current fiber is the home fiber of a thread, so the thread is shutting down
+    RuntimeThreadShutdown(lpFlsData);
+}
+
+REDHAWK_PALEXPORT void REDHAWK_PALAPI PalInitComAndFlsSlot()
+{
+    // Making finalizer thread MTA early ensures that COM is initialized before we initialize our thread
+    // termination callback.
+    CoInitializeEx(NULL, COINIT_MULTITHREADED);
+
+    // We use fiber detach callbacks to run our thread shutdown code because the fiber detach
+    // callback is made without the OS loader lock
+    g_flsIndex = FlsAlloc(FiberDetachCallback);
+    if (g_flsIndex == FLS_OUT_OF_INDEXES)
     {
-        // The current fiber is the home fiber of a thread, so the thread is shutting down
-        RuntimeThreadShutdown(lpFlsData);
+        ASSERT_UNCONDITIONALLY("Could not allocate an FLS slot.");
```

Comment:
Would it be better to return false from the runtime init method when this happens?

Reply:
> Would it be better to return false from the runtime init method when this happens?

We cannot handle a failure to initialize FLS slot, so we have to bail one way or another. 
But, yes, we could handle this failure the same way as an OOM on init path - no failfast, just return from `InitializeGC` with a failure/false.